examples/stream/src/main.rs

```diff
@@ -0,0 +1,140 @@
+                .from_env()?,
+        )
+        .init();
+
+    let maybe_address = std::env::args()
+        .nth(1)
+        .map(|arg| arg.parse::<Multiaddr>())
+        .transpose()
+        .context("Failed to parse argument as `Multiaddr`")?;
+
+    let mut swarm = libp2p::SwarmBuilder::with_new_identity()
+        .with_tokio()
+        .with_quic()
+        .with_behaviour(|_| stream::Behaviour::default())?
+        .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(10)))
+        .build();
+
+    swarm.listen_on("/ip4/127.0.0.1/udp/0/quic-v1".parse()?)?;
+
+    // Register our custom protocol with the behaviour.
+    let (control, mut incoming_streams) = swarm.behaviour_mut().register(PROTOCOL).unwrap();
+
+    // Deal with incoming streams.
+    // Spawning a dedicated task is just one way of doing this.
+    // libp2p doesn't care how you handle incoming streams but you _must_ handle them somehow.
+    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep in processing them.
+    tokio::spawn(async move {
+        // This loop handles incoming streams _sequentially_ but that doesn't have to be the case.
+        // You can also spawn a dedicated task per stream if you want to.
+        // Be aware that this breaks backpressure though as spawning new tasks is equivalent to an unbounded buffer.
+        // Each task needs memory meaning an aggressive remote peer may force you OOM this way.
+
+        while let Some((peer, stream)) = incoming_streams.next().await {
+            if let Err(e) = inbound_ping(stream).await {
+                tracing::warn!(%peer, "Ping protocol failed: {e}");
+                continue;
+            }
+
+            tracing::info!(%peer, "Handled inbound ping!");
+        }
+    });
+
+    // In this demo application, the dialing peer initiates the protocol.
+    if let Some(address) = maybe_address {
+        let Some(Protocol::P2p(peer_id)) = address.iter().last() else {
+            anyhow::bail!("Provided address does not end in `/p2p`");
+        };
+
+        swarm.dial(address)?;
+
+        tokio::spawn(connection_handler(peer_id, control));
+    }
+
+    // Poll the swarm to make progress.
+    loop {
+        let event = swarm.next().await.expect("never terminates");
+
+        match event {
+            libp2p::swarm::SwarmEvent::NewListenAddr { address, .. } => {
+                let listen_address = address.with_p2p(*swarm.local_peer_id()).unwrap();
+                tracing::info!(%listen_address);
+            }
+            event => tracing::trace!(?event),
+        }
+    }
+}
+
+/// A very simple, `async fn`-based connection handler for our custom ping protocol.
+async fn connection_handler(peer: PeerId, mut control: stream::Control) {
+    let mut peer_control = control.peer(peer).await.unwrap();
+
+    loop {
+        tokio::time::sleep(Duration::from_secs(1)).await; // Wait a second between pings.
```

Comment:
Could this sleep as well be moved to the bottom of the `loop`? Or is it necessary to wait for the dial. Moving it to the bottom in my eyes makes this example more focused.

Reply:
Then all `continue` statements would immediately try again. I wanted the delay be effective for those as well.